### PR TITLE
Enhance the fracture generation feature of `mesh_doctor`

### LINKED ISSUE #2857: Improve the fracture generation in `mesh_doctor`
The current fracture generation feature in `mesh_doctor` has limitations:
- All the intersection types between matrices
- Support `VTK_POLYHEDRON`

### PR DESCRIPTION
**Pull request by TotoGaz**
- Complete rewriting of the fracture generation algorithm. There should be not limitation anymore, and it supports `VTK_POLYHEDRON` thanks to @IsaacJu-debug 's investigations.
- Efforts were paid to validate the code using `mypy` and type hints. Not perfect yet, but still a good improvement.

### CODE REVIEW

File: src/coreComponents/python/modules/geosx_mesh_doctor/checks/generate_fractures.py
Comment (TotoGaz) [inline src/coreComponents/python/modules/geosx_mesh_doctor/checks/generate_fractures.py:1]: Note that this is a complete rewriting of the algorithm; therefore it's not particularly relevant to compare the two files during the code review process.

File: src/coreComponents/python/modules/geosx_mesh_doctor/checks/check_fractures.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+def format_collocated_nodes(fracture_mesh: vtkUnstructuredGrid) -> Sequence[Iterable[int]]:
+    """
+    Extract the collocated nodes information from the mesh and formats it in a python way.
+    :param fracture_mesh: The mesh of the fracture (with 2d cells).
+    :return: An iterable over all the buckets of collocated nodes.
+    """
+    collocated_nodes: numpy.ndarray = vtk_to_numpy(fracture_mesh.GetPointData().GetArray("collocated_nodes"))
+    if len(collocated_nodes.shape) == 1:
+        collocated_nodes: numpy.ndarray = collocated_nodes.reshape((collocated_nodes.shape[0], 1))
+    return tuple(map(lambda bucket: tuple(sorted(filter(lambda i: i != -1, bucket))), collocated_nodes))
```
Comment (cssherman) [inline src/coreComponents/python/modules/geosx_mesh_doctor/checks/check_fractures.py]: This is completely fine (and maybe a bit faster?), but I find list comprehension+slicing to be much more readable/pythonic than lambdas+maps+filters:

`tuple([tuple(sorted(bucket[bucket != -1])) for bucket in collected_nodes])`

**Review by cssherman — APPROVED** [review]
Looks good to me

File: src/coreComponents/python/modules/geosx_mesh_doctor/checks/check_fractures.py (hunk 1 above)
Comment (TotoGaz) [reply src/coreComponents/python/modules/geosx_mesh_doctor/checks/check_fractures.py]: You are right, it's far better!

**Pull request merged**